src/native_mux_impl/stdio.rs

```diff
@@ -28,25 +29,74 @@ fn get_null_fd() -> Result<RawFd, Error> {
 }
 
 pub(crate) enum Fd {
-    PipeReadEnd(PipeRead),
-    PipeWriteEnd(PipeWrite),
-
+    Owned(OwnedFd),
     Borrowed(RawFd),
     Null,
 }
 
+fn cvt(ret: c_int) -> io::Result<c_int> {
+    if ret == -1 {
+        Err(io::Error::last_os_error())
+    } else {
+        Ok(ret)
+    }
+}
+
+fn set_blocking(fd: RawFd) -> io::Result<()> {
+    let flags = cvt(unsafe { fcntl(fd, F_GETFL) })?;
+    cvt(unsafe { fcntl(fd, F_SETFL, flags & (!O_NONBLOCK)) })?;
+
+    Ok(())
+}
+
 impl Fd {
     pub(crate) fn as_raw_fd_or_null_fd(&self) -> Result<RawFd, Error> {
         use Fd::*;
 
         match self {
-            PipeReadEnd(fd) => Ok(AsRawFd::as_raw_fd(fd)),
-            PipeWriteEnd(fd) => Ok(AsRawFd::as_raw_fd(fd)),
-
+            Owned(owned_fd) => Ok(owned_fd.as_raw_fd()),
             Borrowed(rawfd) => Ok(*rawfd),
             Null => get_null_fd(),
         }
     }
+
+    /// # Safety
+    ///
+    /// `T::into_raw_fd` must return a valid fd and transfers
+    /// the ownershipt of it.
+    unsafe fn new_owned<T: IntoRawFd>(fd: T) -> Result<Self, Error> {
+        let raw_fd = fd.into_raw_fd();
+        // Create owned_fd so that the fd will be closed on error
+        let owned_fd = OwnedFd::from_raw_fd(raw_fd);
+
+        set_blocking(raw_fd).map_err(Error::ChildIo)?;
```

Comment:
I still think we should keep this call to just before we _need_ it to be blocking, rather than the moment we get one. For example, as written, I _think_ this would also set blocking on the ends of the pipe that we end up keeping? I think it'd be better for this call to go in `to_stdin`/`to_output`.

Reply:
I've moved all the calls to `to_stdin`/`to_output` and also added doc on why read/write is set to blocking.